modules/cudastereo/src/cuda/stereosgm.cu

```diff
@@ -0,0 +1,2087 @@
+    const int x0 = blockIdx.x * (BLOCK_SIZE - WINDOW_WIDTH + 1) - half_kw;
+    const int y0 = blockIdx.y * LINES_PER_BLOCK;
+
+    for (int i = 0; i < WINDOW_HEIGHT; ++i)
+    {
+        const int x = x0 + tid, y = y0 - half_kh + i;
+        pixel_type value = 0;
+        if (0 <= x && x < src.cols && 0 <= y && y < src.rows)
+        {
+            value = src(y, x);
+        }
+        smem_lines[i][tid] = value;
+    }
+    __syncthreads();
+
+#pragma unroll
+    for (int i = 0; i < LINES_PER_BLOCK; ++i)
+    {
+        if (i + 1 < LINES_PER_BLOCK)
+        {
+            // Load to smem
+            const int x = x0 + tid, y = y0 + half_kh + i + 1;
+            pixel_type value = 0;
+            if (0 <= x && x < src.cols && 0 <= y && y < src.rows)
+            {
+                value = src(y, x);
+            }
+            const int smem_x = tid;
+            const int smem_y = (WINDOW_HEIGHT + i) % SMEM_BUFFER_SIZE;
+            smem_lines[smem_y][smem_x] = value;
+        }
+
+        if (half_kw <= tid && tid < BLOCK_SIZE - half_kw)
+        {
+            // Compute and store
+            const int x = x0 + tid, y = y0 + i;
+            if (half_kw <= x && x < src.cols - half_kw && half_kh <= y && y < src.rows - half_kh)
+            {
+                const int smem_x = tid;
+                const int smem_y = (half_kh + i) % SMEM_BUFFER_SIZE;
+                int32_t f = 0;
+                for (int dy = -half_kh; dy < 0; ++dy)
+                {
+                    const int smem_y1 = (smem_y + dy + SMEM_BUFFER_SIZE) % SMEM_BUFFER_SIZE;
+                    const int smem_y2 = (smem_y - dy + SMEM_BUFFER_SIZE) % SMEM_BUFFER_SIZE;
+                    for (int dx = -half_kw; dx <= half_kw; ++dx)
+                    {
+                        const int smem_x1 = smem_x + dx;
+                        const int smem_x2 = smem_x - dx;
+                        const auto a = smem_lines[smem_y1][smem_x1];
+                        const auto b = smem_lines[smem_y2][smem_x2];
+                        f = (f << 1) | (a > b);
+                    }
+                }
+                for (int dx = -half_kw; dx < 0; ++dx)
+                {
+                    const int smem_x1 = smem_x + dx;
+                    const int smem_x2 = smem_x - dx;
+                    const auto a = smem_lines[smem_y][smem_x1];
+                    const auto b = smem_lines[smem_y][smem_x2];
+                    f = (f << 1) | (a > b);
+                }
+                dest(y, x) = f;
+            }
+        }
+        __syncthreads();
+    }
+}
+} // anonymous namespace
+
+void censusTransform(const GpuMat& src, GpuMat& dest, Stream& _stream)
+{
+    CV_Assert(src.size() == dest.size());
+    CV_Assert(src.type() == CV_8UC1 || src.type() == CV_16UC1);
+    const int width_per_block = BLOCK_SIZE - WINDOW_WIDTH + 1;
+    const int height_per_block = LINES_PER_BLOCK;
+    const dim3 gdim(
+        (src.cols + width_per_block - 1) / width_per_block,
```

Comment:
When rounding up integer division in calculating grid dimension, it would be better to use `divUp()`.